iscsi/initiator.go

```
@@ -313,43 +310,50 @@ func findScsiDevice(ip, target string, lun int, ne *util.NamespaceExecutor) (*ut
 		return nil, fmt.Errorf("cannot find kernel device for iSCSI device: %s", name)
 	}
 
-	return dev, nil
+	return &dev, nil
 }
 
-func CleanupScsiNodes(target string, ne *util.NamespaceExecutor) error {
+func CleanupScsiNodes(target string) error {
 	for _, dir := range ScsiNodesDirs {
-		if _, err := ne.Execute("ls", []string{dir}); err != nil {
+		if _, err := lhns.GetFileInfo(dir); err != nil {
 			continue
 		}
+
 		targetDir := filepath.Join(dir, target)
-		if _, err := ne.Execute("ls", []string{targetDir}); err != nil {
+		if _, err := lhns.GetFileInfo(targetDir); err != nil {
 			continue
 		}
+
 		// Remove all empty files in the directory
-		output, err := ne.Execute("find", []string{targetDir})
+		emptyFilePaths, err := lhns.GetEmptyFiles(targetDir)
 		if err != nil {
-			return errors.Wrapf(err, "failed to search iSCSI directory %v", targetDir)
+			return err
 		}
-		scanner := bufio.NewScanner(strings.NewReader(output))
-		for scanner.Scan() {
-			file := scanner.Text()
-			output, err := ne.Execute("stat", []string{file})
+
+		for _, emptyFilePath := range emptyFilePaths {
+			err := lhns.DeletePath(emptyFilePath)
 			if err != nil {
-				return errors.Wrapf(err, "failed to check iSCSI node file %v", file)
+				return errors.Wrapf(err, "failed to clean up empty iSCSI node file %v", emptyFilePath)
 			}
-			if strings.Contains(output, "regular empty file") {
-				if _, err := ne.Execute("rm", []string{file}); err != nil {
-					return errors.Wrapf(err, "failed to clean up empty iSCSI node file %v", file)
-				}
-				// We're trying to clean up the upper level directory as well, but won't mind if we fail
-				_, _ = ne.Execute("rmdir", []string{filepath.Dir(file)})
+		}
+
+		// Try to remove the upper level directory containing empty files.
+		// We don't mind if it fails.
+		dirContainEmptyFiles := make(map[string]bool)
+		for _, emptyFilePath := range emptyFilePaths {
+			dirContainEmptyFiles[filepath.Dir(emptyFilePath)] = true
```

Good to have a log. Don't swallow any error.